### ISSUE TITLE
How to use this library properly?

### ISSUE THREAD

**Issue by KennyThinh** [KevinMusgrave/pytorch-metric-learning#350]
First,  thank you for very helpful library. 

I have gone through your library for a while and still find hard to find whether it is suitable for my current problem or not. So I would like to ask you for advice as well as some instructions.

First, my current problem is to build a face recognition (not for human being, but for dogs). After researching a while, I know that what I need is deep metrics learning, that is why I came up with your library. Currently, my approach is

1. Training phase: Fine tuning resnet, densenet with the last layer outputs 2048 embeddings
2. Inference phase: 
    2.1 Learn embeddings from a gallery: each ID has some pics. I will learn the embeddings for each ID from these pics.
    2.2 Predict a query image by comparing its embedding with the embeddings of each ID in the gallery.
    For the inference phase, I have used 2 ways:
        (1) Use Inference_model to fit on gallery data and use get_nearest_neighbors to find the most similar to the query image
        (2) Combining embeddings of pics of each ID in gallery, and compare the combined embeddings with the one of query image. Unluckily, these 2 ways are not working well. So I got doubt about whether my approach is right or not, as well as whether I am using your library properly or not. So I would like to ask you that

(1) The above approach is the right approach to face recognition problem?
(2) Maybe my bad result are due to bad model training. I have follow exactly your example `TripletMarginLossMNIST.ipynb` to train model. But the loss curve was very fluctuated (may times the loss was 0, and not going down after epochs). You also said some where that the loss  =0 is very normal. If so, the loss curse fluctuated (attached image) is normal?
![image](https://user-images.githubusercontent.com/16069257/123977464-f8d93280-d9f9-11eb-8e62-508386a5f1e3.png)
(3) In `TripletMarginLossMNIST.ipynb` we trained model by writing your own train function. Does trainer module do the same thing?
(4) What is the proper metrics should I use to evaluate the model? The loss on validation set? or precision_at_1 at validation set? When measure loss of validation set, do we need to do exactly as when we train, i.e, 
`embeddings = model(data)
indices_tuple = mining_func(embeddings, labels)
loss = loss_func(embeddings, labels, indices_tuple)
loss = loss_func(embeddings, labels)`
(5) Can you suggest a good approach to use your library for my problem? e.g
- which notebook example is the closest to my problem?
 - which kind of loss to try first?
 - Do I need embedder and embedder optimizer? 
  
I know there are a lot of question. But hope that you can give me some hints to keep me dig deeper to your library. Thanks you very much.
>

**Comment by KevinMusgrave** [comment]
> (1) The above approach is the right approach to face recognition problem?

It seems reasonable

> (2) Maybe my bad result are due to bad model training. I have follow exactly your example `TripletMarginLossMNIST.ipynb` to train model. But the loss curve was very fluctuated (may times the loss was 0, and not going down after epochs). You also said some where that the loss =0 is very normal. If so, the loss curse fluctuated (attached image) is normal?
> ![image](https://user-images.githubusercontent.com/16069257/123977464-f8d93280-d9f9-11eb-8e62-508386a5f1e3.png)

It's normal for the triplet loss to stay at a constant value. This is because 0 loss triplets are discarded from the calculation, and so the average at each iteration never goes down.

> (3) In `TripletMarginLossMNIST.ipynb` we trained model by writing your own train function. Does trainer module do the same thing?

Yes, you can check out the MetricLossOnly notebook in the same examples folder.

> (4) What is the proper metrics should I use to evaluate the model? 

Precision at 1 on the validation set will be a good indicator of how well the model is doing. 

> (5) Can you suggest a good approach to use your library for my problem? e.g

I would try ContrastiveLoss with no miners. You may have to adjust the ```pos_margin``` and ```neg_margin``` to improve results. Make sure the batch size is at least 32.

You can also try ```torch.nn.CrossEntropyLoss```. In this case, you need to append ```nn.Linear(2048, num_classes)``` to your network. Then during inference, use the 2048 layer as the embeddings.

**Comment by KennyThinh** [comment]
Sorry for my late reply. Thank you very much.
